pkgs/development/compilers/rust/rustc.nix

```diff
@@ -145,14 +146,19 @@ in stdenv.mkDerivation (finalAttrs: {
     # Since fastCross only builds std, it doesn't make sense (and
     # doesn't work) to build a linker.
     "--disable-llvm-bitcode-linker"
-  ] ++ optionals (stdenv.isLinux && !stdenv.targetPlatform.isRedox) [
+  ] ++ optionals (stdenv.isLinux && !stdenv.targetPlatform.isRedox && !stdenv.targetPlatform.isWasi) [
     "--enable-profiler" # build libprofiler_builtins
   ] ++ optionals stdenv.buildPlatform.isMusl [
     "${setBuild}.musl-root=${pkgsBuildBuild.targetPackages.stdenv.cc.libc}"
   ] ++ optionals stdenv.hostPlatform.isMusl [
     "${setHost}.musl-root=${pkgsBuildHost.targetPackages.stdenv.cc.libc}"
   ] ++ optionals stdenv.targetPlatform.isMusl [
     "${setTarget}.musl-root=${pkgsBuildTarget.targetPackages.stdenv.cc.libc}"
+  ] ++ optionals stdenv.targetPlatform.isWasi [
+    "${setTarget}.wasi-root=${runCommandLocal "wasi-sysroot" {} ''
+      mkdir -p $out/lib
+      ln -s ${pkgsBuildTarget.targetPackages.stdenv.cc.libc}/lib $out/lib/wasm32-wasi
+    ''}"
```

Comment:
This workaround is required because wasilibc [overrides](https://github.com/NixOS/nixpkgs/blob/8fa00a61925cf7fe2c302b8d0d9bb5119a7388bb/pkgs/development/libraries/wasilibc/default.nix#L33) the default path/folder structure it's being installed to, but rust expects that structure to exist. The rustc build copies the files from that folder, so this derivation won't get referenced in the closure.

Alternatively this could be done by
1. creating a symlink with the right folder stucture in wasilibc's postInstall (which isn't pretty either and would cause a rebuild of firefox)
2. creating some tmp directory with the right folder structure during rustc's postPatch (but that would mean this workaround must be spread over two different places in this file, which I'd prefer to avoid.

So I think this is the prettiest way of handling that.

Reply:
> creating a symlink with the right folder stucture in wasilibc's postInstall like [so](https://github.com/NixOS/nixpkgs/compare/8aa5aa70d620106e0be3cab003d5cd0c49fae874..d57e347cee6e74743e019b0b30bb2a3230837b96) (which isn't pretty either and would mean I have to touch something that's in the build path of firefox, which I'd like to avoid)

I think this makes more sense.  If Rust expects it here, other things might too.  This is not an urgent change, so there's definitely no reason avoiding a one-off Firefox rebuild should be a consideration.

Reply:
If you say "other things too", I should probably create that symlink for the headers, too. Done.